File: grip_core/utils/message_utils.cpp
/*
* Copyright 2020, 2023 Shadow Robot Company Ltd.
*
* This program is free software: you can redistribute it and/or modify it
* under the terms of the GNU General Public License as published by the Free
* Software Foundation version 2 of the License.
*
* This program is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
* more details.
*
* You should have received a copy of the GNU General Public License along
* with this program. If not, see <http://www.gnu.org/licenses/>.
*/

#include <vector>
#include <string>
#include <grip_core/message_utils.hpp>

geometry_msgs::PoseStamped generate_pose_stamped_message(std::string reference_frame_name, Eigen::Vector3f position,
                                                         Eigen::Vector3f orientation)
{
    geometry_msgs::PoseStamped message_to_return;
    tf2::Quaternion quaternion;
    message_to_return.header.frame_id = reference_frame_name;
    message_to_return.header.stamp = ros::Time::now();
    message_to_return.pose.position.x = position[0];
    message_to_return.pose.position.y = position[1];
    message_to_return.pose.position.z = position[2];
    quaternion.setRPY(orientation[0], orientation[1], orientation[2]);
    message_to_return.pose.orientation.x = quaternion.x();
    message_to_return.pose.orientation.y = quaternion.y();
    message_to_return.pose.orientation.z = quaternion.z();
    message_to_return.pose.orientation.w = quaternion.w();
    return message_to_return;
}

geometry_msgs::PoseStamped generate_pose_stamped_message(std::string reference_frame_name, Eigen::Vector3f position,
                                                         Eigen::Vector4f orientation)
{
    geometry_msgs::PoseStamped message_to_return;
    tf2::Quaternion quaternion;
    message_to_return.header.frame_id = reference_frame_name;
    message_to_return.header.stamp = ros::Time::now();
    message_to_return.pose.position.x = position[0];
    message_to_return.pose.position.y = position[1];
    message_to_return.pose.position.z = position[2];
    message_to_return.pose.orientation.x = orientation[0];
    message_to_return.pose.orientation.y = orientation[1];
    message_to_return.pose.orientation.z = orientation[2];
    message_to_return.pose.orientation.w = orientation[3];
    return message_to_return;
}

sensor_msgs::JointState generate_joint_state_message(std::vector<std::string> joint_names, std::vector<double> position,
                                                     std::vector<double> velocity, std::vector<double> effort,
                                                     std::string reference_frame_name)
{
    sensor_msgs::JointState message_to_return;
    message_to_return.header.frame_id = reference_frame_name;
    message_to_return.header.stamp = ros::Time::now();
    message_to_return.name = joint_names;
    message_to_return.position = position;
    message_to_return.velocity = velocity;
    message_to_return.effort = effort;
    return message_to_return;
}

File: grip_core/src/grip_core/utils/message_utils.py
#!/usr/bin/env python3

# Copyright 2019, 2023 Shadow Robot Company Ltd.
#
# This program is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation version 2 of the License.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

import rospy
from geometry_msgs.msg import PoseStamped
from sensor_msgs.msg import JointState
from tf.transformations import quaternion_from_euler


def generate_pose_stamped_message(reference_frame_name, position, orientation):
    """
        Given an orientation and a position provided in a given reference frame
        return the corresponding PoseStamped message

        @param reference_frame_name: String containing the name of the reference frame
        @param position: List or tuple containing the position (x, y, z)
        @param orientation: List or tuple containing the orientation in RPY system (radians)

        @return: PoseStamped message
    """
    joint_state_to_return = PoseStamped()
    joint_state_to_return.header.frame_id = reference_frame_name
    joint_state_to_return.header.stamp = rospy.Time.now()
    joint_state_to_return.pose.position.x = position[0]
    joint_state_to_return.pose.position.y = position[1]
    joint_state_to_return.pose.position.z = position[2]
    quaternion = quaternion_from_euler(orientation[0], orientation[1], orientation[2])
    joint_state_to_return.pose.orientation.x = quaternion[0]
    joint_state_to_return.pose.orientation.y = quaternion[1]
    joint_state_to_return.pose.orientation.z = quaternion[2]
    joint_state_to_return.pose.orientation.w = quaternion[3]
    return joint_state_to_return


def generate_joint_state_message(joint_names, position, velocity=[], effort=[], reference_frame_name="world"):
    """
        Given the position, velocity (optional) and effort (optional) given in a reference frame
        return the corresponding JointState message

        @param joint_names: List of string stating which joints are refered to
        @param position: List or tuple containing the position (rad or m) of the joints defined in joint_names
        @param position: List or tuple containing the velocity of the joints defined in joint_names
        @param position: List or tuple containing the effort of the joints defined in joint_names
        @param reference_frame_name: String containing the name of the reference frame

        @return: JointState message
    """
    joint_state_to_return = JointState()
    joint_state_to_return.header.frame_id = reference_frame_name
    joint_state_to_return.header.stamp = rospy.Time.now()
    joint_state_to_return.name = joint_names
    joint_state_to_return.position = position
    joint_state_to_return.velocity = velocity
    joint_state_to_return.effort = effort
    return joint_state_to_return
